subsys/fs/zffs/src/zffs_file.c

```diff
@@ -0,0 +1,456 @@
+
+	rc = zffs_dir_append(zffs, pointer, dir, node_data);
+	if (rc) {
+		return rc;
+	}
+
+	file->id = node_data->id;
+	file->size = 0;
+	file->offset = 0;
+	file->next_id = node_data->file.next_id;
+	zffs_dlist_init(&file->list);
+	file->block.id = ZFFS_NULL;
+	file->node.id = ZFFS_NULL;
+	file->node.prev = ZFFS_NULL;
+	file->node.next = ZFFS_NULL;
+	file->write.block.id = ZFFS_NULL;
+
+	sys_slist_append(&zffs->opened, (void *)file);
+	return 0;
+}
+
+int zffs_file_open(struct zffs_data *zffs,
+		   struct zffs_area_pointer *pointer,
+		   const struct zffs_node_data *node_data,
+		   struct zffs_file *file)
+{
+	int rc;
+
+	if (node_data->type != ZFFS_TYPE_FILE) {
+		return -EISDIR;
+	}
+
+	file->id = node_data->id;
+	file->size = node_data->file.size;
+	file->next_id = node_data->file.next_id;
+	file->offset = 0;
+	file->list.wait_update = false;
+	file->list.head = node_data->file.head;
+	file->list.tail = node_data->file.tail;
+	file->write.block.id = ZFFS_NULL;
+
+	if (zffs_dlist_is_empty(&file->list)) {
+		file->block.id = ZFFS_NULL;
+		file->node.id = ZFFS_NULL;
+		file->node.prev = ZFFS_NULL;
+		file->node.next = ZFFS_NULL;
+	} else {
+		struct file_dlist_node_disk disk;
+
+		rc = zffs_dlist_head(zffs, pointer, &file->list, &file->node);
+		if (rc < 0) {
+			return rc;
+		}
+
+		if (rc < sizeof(disk)) {
+			return -EIO;
+		}
+
+		rc = zffs_area_read(zffs, pointer, &disk, sizeof(disk));
+		if (rc) {
+			return rc;
+		}
+
+		rc = zffs_block_open(zffs, pointer, sys_get_le32(disk.block),
+				     &file->block);
+		if (rc < 0) {
+			return rc;
+		}
+
+		if (rc != ZFFS_CONFIG_BLOCK_SIZE) {
+			return -EIO;
+		}
+
+		file->pointer = *pointer;
+	}
+
+	sys_slist_append(&zffs->opened, (void *)file);
+
+	return 0;
+}
+
+static int file_sync(struct zffs_data *zffs,
+		     struct zffs_area_pointer *pointer,
+		     struct zffs_file *file)
+{
+	int rc;
+
+	if (file_is_write(file)) {
+		struct file_dlist_node_disk disk;
+		zffs_disk(u16_t, crc);
+
+		if (((file->offset & (ZFFS_CONFIG_BLOCK_SIZE - 1)) != 0)) {
+			struct zffs_area_pointer rp = file->pointer;
+			u32_t copybytes = file_write_free_size(file);
+
+			file->pointer = file->write.pointer;
+
+			rc = file->block.id == ZFFS_NULL
+			     ? zffs_area_crc(zffs, &file->write.pointer,
+					     copybytes, &file->write.crc)
+			     : zffs_area_copy_crc(
+				zffs, &rp, &file->write.pointer,
+				copybytes, &file->write.crc);
+			if (rc) {
+				return rc;
+			}
+		}
+
+		sys_put_le16(file->write.crc, crc);
+
+		rc = zffs_area_write(zffs, &file->write.pointer, crc,
+				     sizeof(crc));
+		if (rc) {
+			return rc;
+		}
+
+		sys_put_le32(file->write.block.id, disk.block);
+
+		if (file->block.id == ZFFS_NULL) {
+			file->node.id =
+				zffs_misc_next_id(zffs, &file->next_id);
+
+			rc = zffs_dlist_append(zffs, pointer, &file->list,
+					       &file->node, &disk,
+					       sizeof(disk));
+		} else {
+			rc = zffs_dlist_updata_ex(
+				zffs, pointer, file->node.id, &disk, sizeof(disk));
```

Comment:
you could have put zffs, pointer in the beginning of parameter list, etc...

Reply:
yep